### PR TITLE
21 batch size and collation

### LINKED ISSUE #21: Batch size and Collation
**Acceptance Criteria**:
- batch size is configurable (via CLI or config files if those are set up already)
- batches are collated by padding the end of shorter token tensors. Attention masks mask out the padding.

### PR DESCRIPTION
**Pull request by JansenNye**
Ran training tests on wikitext dataset using batch sizes of 1 and 2. 
Closes #21

### CODE REVIEW

File: src/mirror/util.py [hunk 1]
```diff
@@ -21,3 +22,20 @@ def assert_can_download(item_name_to_download: str):
 
 def is_power_of_ten(n: int):
     return n > 0 and math.log10(n).is_integer()
+
+def pad_to_longest(batch, pad_token):
```
Comment (Plyb) [inline src/mirror/util.py]: can we put some type hints on this? there's a types file that might have a type that would work for the batch.

File: src/mirror/util.py [hunk 2]
```diff
@@ -21,3 +22,20 @@ def assert_can_download(item_name_to_download: str):
...
+    device = batch[0].device if torch.is_tensor(batch[0]) else torch.device("cpu")
+    lens = torch.tensor([b.numel() for b in batch], device=device, dtype=torch.long)
+    max_len = int(lens.max().item())
+    batch_size = len(batch)
+
+    tokens = torch.full((batch_size, max_len), int(pad_token), dtype=torch.long, device=device)
+
+    ar = torch.arange(max_len, device=batch[0].device)
+    attention_mask = (ar.unsqueeze(0) < lens.unsqueeze(1)).to(torch.long)
```
Comment (Plyb) [inline src/mirror/util.py]: Nifty! It might be worth running a test printing out the attention masks just to make sure this is doing what we expect it to.

**Comment by JansenNye** [comment]
Ran training tests on wikitext dataset using batch sizes of 1 and 2 as well as randomized examples from the placeholder model

File: src/mirror/models/placeholder_model.py [hunk 3]
```diff
@@ -11,8 +11,8 @@ class PlaceholderModel(MirrorModel):
     def __init__(self) -> None:
         super().__init__()
         self.parameter = nn.Parameter(torch.tensor([0.0], device=device))
-
-    _tokenizer = PlaceholderTokenizer()
+        self.pad_token_id = -1
```
[the remark this answers is not in the document]
Comment (Plyb) [reply src/mirror/models/placeholder_model.py]: This still needs to be removed

File: src/mirror/util.py (hunk 1 above)
Comment (Plyb) [reply src/mirror/util.py]: This also still needs to be done

File: src/mirror/util.py [hunk 4]
```diff
@@ -21,3 +22,26 @@ def assert_can_download(item_name_to_download: str):
...
+    batch_size = len(batch)
+
+    tokens = torch.full((batch_size, max_len), int(pad_token), dtype=torch.long, device=dev)
+
+    ar = torch.arange(max_len, device=dev)
+    attention_mask = (ar.unsqueeze(0) < lens.unsqueeze(1)).to(torch.long)
+
+    print("raw batch[0]:", batch[0])
+    print("lens:", lens)
+    print("attention_mask:\n", attention_mask)
```
Comment (Plyb) [inline src/mirror/util.py]: These should be removed please

**Review by Plyb — CHANGES_REQUESTED** [review]
Looks like some comments didn't get addressed. Maybe you just forgot to push?

File: src/mirror/util.py [hunk 5]
```diff
@@ -21,3 +23,20 @@ def assert_can_download(item_name_to_download: str):
...
 def is_power_of_ten(n: int):
     return n > 0 and math.log10(n).is_integer()
+
+def pad_to_longest(batch: TokenBatch, pad_token: int):
+    dev = device
+    lens = torch.tensor([b.numel() for b in batch], device=dev, dtype=torch.long)
+    max_len = lens.max().item()
+    batch_size = len(batch)
+
+    tokens = torch.full((batch_size, max_len), int(pad_token), dtype=torch.long, device=dev)
```
Comment (Plyb) [inline src/mirror/util.py]: is there any particular reason to not just use `device` directly on line 29 and line 33? as opposed to making a new variable for it

**Review by Plyb — CHANGES_REQUESTED** [review]
last thing
